src/main/java/com/yelp/nrtsearch/server/luceneserver/nrt/ProportionalCopyThread.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2024 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.nrt;
+
+import java.util.PriorityQueue;
+import org.apache.lucene.replicator.nrt.CopyJob;
+import org.apache.lucene.replicator.nrt.Node;
+
+/**
+ * Copy thread that uses separate priority queues for high and low priority jobs. The value of
+ * lowPriorityCopyPercentage determines the proportion of times a low priority job is selected over
+ * a high priority job. For example, if lowPriorityCopyPercentage is 10, then a low priority job
+ * will be selected 10% of the time and a high priority job will be selected 90% of the time. If
+ * there are no high priority jobs, then a low priority job will always be selected.
+ */
+public class ProportionalCopyThread extends NrtCopyThread {
+  private final PriorityQueue<CopyJob> highPriorityQueue = new PriorityQueue<>();
+  private final PriorityQueue<CopyJob> lowPriorityQueue = new PriorityQueue<>();
+  private final int lowPriorityCopyPercentage;
+  private long counter = 0;
+
+  public ProportionalCopyThread(Node node, int lowPriorityCopyPercentage) {
+    super(node);
+    if (lowPriorityCopyPercentage < 0 || lowPriorityCopyPercentage > 100) {
+      throw new IllegalArgumentException(
+          "lowPriorityCopyPercentage must be between 0 and 100, inclusive. Got: "
+              + lowPriorityCopyPercentage);
+    }
+    this.lowPriorityCopyPercentage = lowPriorityCopyPercentage;
+  }
+
+  @Override
+  boolean hasJob() {
+    return !highPriorityQueue.isEmpty() || !lowPriorityQueue.isEmpty();
+  }
+
+  @Override
+  CopyJob getJob() {
+    CopyJob job;
+    if (counter++ % 100 < lowPriorityCopyPercentage) {
```

Comment:
Just curious and wanted to check couple of things:

1.  when counter starts from 0 (say 0 to 10) it would pick from low priority queue even if there are tasks in higher priority queue.  But I assume its okay given large number of tasks.
2.  Am I right that when counter hits Long.MAX_VALUE `counter++ % 100` will be a negative value for a long time and would satisfy the first condition.

Reply:
I'm not sure 1 is a big deal. Initially it will process whatever job arrives first. It is unknown where the counter will be when one of each priority job is available. Another option would be to use random numbers, but I'm not sure it makes a big difference, and the batching might actually be better for acked transfers.

For 2, I don't think we need to worry about an overflow. The maximum long value is 9,223,372,036,854,775,807. Even if we incremented 1M a second, it would take 100K+ years to overflow.